src/core/lib/surface/call.c

```diff
@@ -1384,8 +1397,8 @@ static void receiving_initial_metadata_ready(grpc_exec_ctx *exec_ctx,
     }
   }
 
-  call->has_initial_md_been_received = true;
-  if (call->saved_receiving_stream_ready_bctlp != NULL) {
+  if (!gpr_atm_acq_cas(&call->recv_state, RECV_NONE,
```

Comment:
I've checked the previous implementation in 873bb70, it should not have the issue with `__ATOMIC_RELAXED` semantics in failed `cas`. It's a little more complicated but does not require new `gpr_atm_xxx_cas` APIs.